cmd/lock-rpc-server.go

```diff
@@ -158,26 +169,134 @@ func (l *lockServer) RUnlock(args *LockArgs, reply *bool) error {
+				*lri = append((*lri)[:index], (*lri)[index+1:]...)
+				l.lockMap[name] = *lri
+			}
+			return true
+		}
+	}
+	return false
+}
+
+// nameLockRequesterInfoPair is a helper type for lock maintenance
+type nameLockRequesterInfoPair struct {
+	name string
+	lri  lockRequesterInfo
+}
+
+// getLongLivedLocks returns locks that are older than a certain time and
+// have not been 'checked' for validity too soon enough
+func getLongLivedLocks(m map[string][]lockRequesterInfo, interval time.Duration) []nameLockRequesterInfoPair {
+
+	rslt := []nameLockRequesterInfoPair{}
+
+	for name, lriArray := range m {
+
+		for idx := range lriArray {
+			// Check whether enough time has gone by since last check
+			if time.Since(lriArray[idx].timeLastCheck) >= interval {
+				rslt = append(rslt, nameLockRequesterInfoPair{name: name, lri: lriArray[idx]})
+				lriArray[idx].timeLastCheck = time.Now()
+			}
+		}
+	}
+
+	return rslt
+}
+
+// lockMaintenance loops over locks that have been active for some time and checks back
+// with the original server whether it is still alive or not
+func (l *lockServer) lockMaintenance(interval time.Duration) {
+
+	l.mutex.Lock()
+	// get list of locks to check
+	nlripLongLived := getLongLivedLocks(l.lockMap, interval)
+	l.mutex.Unlock()
+
+	for _, nlrip := range nlripLongLived {
+
+		c := newClient(nlrip.lri.node, nlrip.lri.rpcPath)
+
+		var active bool
+
+		// Call back to original server verify whether the lock is still active (based on name & uid)
+		if err := c.Call("Dsync.Active", &LockArgs{Name: nlrip.name, UID: nlrip.lri.uid}, &active); err != nil {
+			// We failed to connect back to the server that originated the lock, this can either be due to
+			// - server at client down
+			// - some network error (and server is up normally)
+			//
+			// We will ignore the error, and we will retry later to get resolve on this lock
+			log.Println("  Dsync.Active failed:", err)
+			c.Close()
+		} else {
+			c.Close()
+			log.Println("  Dsync.Active result:", active)
+
+			if !active { // The lock is no longer active at server that originated the lock
+				// so remove the lock from the map
+				l.mutex.Lock()
+				// Check if entry is still in map (could have been removed altogether by 'concurrent' (R)Unlock of last entry)
+				if lri, ok := l.lockMap[nlrip.name]; ok {
+					if !l.removeEntry(nlrip.name, nlrip.lri.uid, &lri) {
+						// Remove failed, in case it is a:
+						if nlrip.lri.writer {
+							// Writer: this should never happen as the whole (mapped) entry should have been deleted
+							log.Println("Lock maintenance failed to remove entry for write lock (should never happen)", nlrip.name, nlrip.lri, lri)
+						} else {
+							// Reader: this can happen if multiple read locks were active and the one we are looking for
+							// has been released concurrently (so it is fine)
+						}
+					} else {
+						// remove went okay, all is fine
+					}
+				}
+				l.mutex.Unlock()
+			}
+		}
+	}
```

Comment:
That would be nice, but we'd have to think about how to do this (you would have to fire up multiple 'minio server's on localhost, perform operations on it so that distributed locks are used and then kill servers and wait for maintenance to occur, and retry again).